llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -4605,6 +4605,30 @@ static Instruction *foldICmpXorXX(ICmpInst &I, const SimplifyQuery &Q,
       isKnownNonZero(A, Q.DL, /*Depth=*/0, Q.AC, Q.CxtI, Q.DT))
     return new ICmpInst(PredOut, Op0, Op1);
 
+  // These transform works when C is negative.
+  // X s< X^C, X s<= X^C, X u> X^C, X u>= X^C  --> X s< 0
+  // X s> X^C, X s>= X^C, X u< X^C, X u<= X^C  --> X s>= 0
+  if (auto C = dyn_cast<Constant>(A)) {
+    if (!C->getUniqueInteger().isNegative())
+      return nullptr;
```

Comment:
This should use `m_Negative()` instead. getUniqueInteger() will assert on non-splat integer. Should have one test case for this.